### ISSUE TITLE
Make it possible to override RangeX 

### ISSUE THREAD

**Issue by Delphi-FPC-Lazarus** [JAM-Software/Virtual-TreeView#670]
FRangeX is calculated by TBaseVirtualTree.UpdateHorizontalRange() from header FHeader.FColumns.TotalWidth.

It would be useful if it can be overwritten in some special cases when columns width get's larger than it's header, such as a subtable.
![subtable](https://cloud.githubusercontent.com/assets/23053293/19684753/52d31aba-9ab9-11e6-812d-dbe1bce497d2.png)

In this case it is inpossible to scroll to the end of the subtable because FRangeX / HorzScrollbar is calculated by header width.

**I guess to make a few little changes in VirtualTrees.pas to make it possible:**

`+    procedure SetRangeX(value: Cardinal);`

`-    property RangeX: Cardinal read GetRangeX;`
`+    property RangeX: Cardinal read GetRangeX write SetRangeX;`

`-    procedure UpdateHorizontalRange;`
`+    procedure UpdateHorizontalRange; virtual;`

`+procedure TBaseVirtualTree.SetRangeX(value: Cardinal);`
`+begin`
`+  FRangeX := value;`
`+end;`

---

With this changes, UpdateHorizontalRange can be overwritten,
after calling inherited, RangeX can be modified for special cases like subtables.

regards
Peter

**Comment by joachimmarder** [comment]
There is a protected SetRangeX() now and UpdateHorizontalRange is virtual.

**Comment by Delphi-FPC-Lazarus** [comment]
thankyou :)

regards
Peter



Am 29.11.2016 um 20:58 schrieb Joachim Marder:

**Comment by jovanvasic** [comment]
Hi there, I have found a side effect with this change request.  Taking Peter's example to adjust the RangeX value to accommodate the subtable breaks the function TVTHeader.DetermineSplitterIndex. RangeX value is used to initialize the Variable SplitPoint for the determination of the column splitter.  If I haven't missed anything else following change should fix the problem.
`- SplitPoint := -Treeview.FEffectiveOffsetX + Integer(Treeview.FRangeX);`
`+ SplitPoint := -Treeview.FEffectiveOffsetX + Integer(FColumns.TotalWidth);`

**Comment by jovanvasic** [comment]
And to pick up on my previous post and complete my request... I have missed the block TreeView.UseRightToLeftAlignment in the function TVTHeader.DetermineSplitterIndex.
`- if Integer(Treeview.FRangeX) < Treeview.ClientWidth then`
`+ if Integer(FColumns.TotalWidth) < Treeview.ClientWidth then`

`-    Inc(SplitPoint, Treeview.ClientWidth - Integer(Treeview.FRangeX));`
`+    Inc(SplitPoint, Treeview.ClientWidth - Integer(FColumns.TotalWidth));`

**Comment by joachimmarder** [comment]
@jovanvasic: It's difficult to judge your proposed changes without having a sample project that makes use of overriding `SetRangeX()` and `UpdateHorizontalRange()`, the use of this seems rather specific.

@mp3freak: Can you comment on them?

**Comment by jovanvasic** [comment]
Hi Joachim, I attached a sample project to show what I'm trying to do. On a click in the first column of a node the node expands and displays a sub table. The sub table is a TVirtualStringTree and it is wider then the main table. The Scrollbar in the sub table is deactivated. Using your announced changes for UpdateHorizontalRange() and SetRangeX() I manipulate the RangeX value of the main table with the RangeX value from the sub table. This allows me to adjust the Scroll Range to access the content of the sub table using the main table scrollbar. Now to come back to yesterdays post, if a node in the main table is expanded and the main table RangeX value therefore manipulated the function TVTHeader.DetermineSplitterIndex cannot determine the SplitterIndex anymore because it uses the RangeX value. Testing and comparing the values I found the RangeX value is the same like the Columns.TotalWidth. Being in a function of the header to find the splitter points I thought it may be better to use the Columns.TotalWidth instead, in case the RangeX gets modified.

[VirtualTreeRangeX_Demo.zip](https://github.com/Virtual-TreeView/Virtual-TreeView/files/702220/VirtualTreeRangeX_Demo.zip)
